### PR TITLE
feat: add module with version update notifications

### PR DESCRIPTION
**Pull request by NikkyAI**
well here is my first attempt
i tried to leave enough of the API public so that users can use `fun RegularAction<*>.fetchAvailableVersions(): List<Version>` themselves

as well as the github logging (notice, warning, error) and `GITHUB_STEP_SUMMARY` append functions

this wil lalso be useful for iterating on this in workflows without needing to push new versions until we find formatting and such that people are happy with

### CODE REVIEW

File: .jitpack.yml
Comment (krzema12) [inline .jitpack.yml:1]: I moved away from JitPack in favor of Maven snapshots. It's fine for testing, but let's not commit it to the repo.

File: .github/workflows/outdated_workflow.main.kts
Comment (krzema12) [inline .github/workflows/outdated_workflow.main.kts:1]: Once we test end-to-ensd that stuff works, let's remove this workflow, and instead have some tests lower-level tests.

File: action-versions-updates/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    kotlin("plugin.serialization")
+
+    id("org.jetbrains.kotlinx.binary-compatibility-validator") version "0.14.0"
+}
+
+group = rootProject.group
+version = rootProject.version
+
+dependencies {
+    // jitpack (or scripting?) seems to have issues when dependencies are coming from a bom
```
Comment (krzema12) [inline action-versions-updates/build.gradle.kts:13]: A note to remember using the bom.

File: action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/internal/Http.kt
Comment (krzema12) [inline action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/internal/Http.kt:1]: It's copied from another module - let's try putting it in a shared place, like shared-internal.

File: action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/ActionVersionUpdates.kt [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+public fun Workflow.getAvailableActionVersionUpdates(): List<VersionUpdateAvailable> {
+    val actionSteps =
+        jobs
+            .flatMap { job ->
+                job.steps
+            }
+            .filterIsInstance<ActionStep<*>>()
+
+    return actionSteps.mapNotNull { step ->
```
Comment (krzema12) [inline action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/ActionVersionUpdates.kt:26]: There are no logs as this logic iterates through the actions, and it takes minutes to run.

File: buildSrc/src/main/kotlin/buildsrc/convention/publishing.gradle.kts [hunk 3]
```diff
@@ -63,8 +63,13 @@ publishing {
     }
 }
 
+val jitpack = System.getenv("JITPACK")?.toString()?.toBooleanStrictOrNull() ?: false
```
Comment (krzema12) [inline buildSrc/src/main/kotlin/buildsrc/convention/publishing.gradle.kts:66]: Let's also revert these changes.

**Review by krzema12 — CHANGES_REQUESTED** [review]
Partial review.

File: action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/ActionVersionUpdates.kt (hunk 2 above)
Comment (krzema12) [reply action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/ActionVersionUpdates.kt:26]: It took over 6 minutes to run it, let's try to make it faster. I wonder why it takes so long.

File: action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/internal/Http.kt
Comment (krzema12) [reply action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/internal/Http.kt:1]: I've extracted
```kotlin
suspend fun fetchAvailableVersions(owner: String,name: String,githubToken: String): List<Version>
```

to `shared-internal` module, let's reuse it.

File: .jitpack.yml
Comment (NikkyAI) [reply .jitpack.yml:1]: agreed

File: .github/workflows/outdated_workflow.main.kts
Comment (NikkyAI) [reply .github/workflows/outdated_workflow.main.kts:1]: sensible, we could make tests that write the summary to a file and analyze the file content instead if needed

File: action-versions-updates/build.gradle.kts (hunk 1 above)
Comment (NikkyAI) [reply action-versions-updates/build.gradle.kts:13]: lets hope this was jitpack causing issues and not kotlin-scripting then

File: action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/ActionVersionUpdates.kt (hunk 2 above)
Comment (NikkyAI) [reply action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/ActionVersionUpdates.kt:26]: it does not take that long locally.. i think most of that was waiting on jitpack before the script managed to compile on github actions ?

Comment (krzema12) [reply action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/ActionVersionUpdates.kt:26]: Ah, that might be it!

File: buildSrc/src/main/kotlin/buildsrc/convention/publishing.gradle.kts (hunk 3 above)
Comment (NikkyAI) [reply buildSrc/src/main/kotlin/buildsrc/convention/publishing.gradle.kts]: some of these are good.. they avoid issues like `Cannot perform signing task ':action-versions-updates:signMavenJavaPublication' because it has no configured signatory`
which happens when you do not have keys set up locally.. lowering the barrier for contributotors

Comment (krzema12) [reply buildSrc/src/main/kotlin/buildsrc/convention/publishing.gradle.kts]: It doesn't belong to this PR, we can discuss in a separate one.

Comment (NikkyAI) [reply buildSrc/src/main/kotlin/buildsrc/convention/publishing.gradle.kts]: okay, then li will extract it into a seperate branch.. it makes testing this really hard when i cannot `publishToMavenLocal` locally or in actions

Comment (NikkyAI) [reply buildSrc/src/main/kotlin/buildsrc/convention/publishing.gradle.kts]: created a separate PR https://github.com/typesafegithub/github-workflows-kt/pull/1402

File: action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/GithubLogging.kt
Comment (krzema12) [inline action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/GithubLogging.kt:1]: Does stuff in this file have to be public? If not, let's make it `internal`.

File: action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/GithubLogging.kt [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+            file = file,
+            col = col,
+            endColumn = endColumn,
+            line = line,
+            endLine = endLine,
+        )
+    }
+}
+
+public fun githubError(
```
Comment (krzema12) [inline action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/GithubLogging.kt:126]: Is this function needed? I don't see it being used.

File: action-versions-updates/src/test/kotlin/io/github/typesafegithub/workflows/dsl/CheckVersionsTests.kt [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+import io.github.typesafegithub.workflows.updates.findDependencyDeclaration
+import io.github.typesafegithub.workflows.updates.mavenCoordinatesForAction
+import io.github.typesafegithub.workflows.updates.reportAvailableUpdates
+import io.kotest.core.spec.style.FunSpec
+import io.kotest.engine.spec.tempdir
+import io.kotest.matchers.collections.shouldHaveSize
+import io.kotest.matchers.shouldBe
+import io.kotest.matchers.string.shouldNotBeEmpty
+
+class CheckVersionsTests : FunSpec(
```
Comment (krzema12) [inline action-versions-updates/src/test/kotlin/io/github/typesafegithub/workflows/dsl/CheckVersionsTests.kt:17]: The tests are failing, see https://github.com/typesafegithub/github-workflows-kt/actions/runs/8797297289/job/24152141997?pr=1393

File: shared-internal/build.gradle.kts [hunk 6]
```diff
@@ -9,9 +9,9 @@ group = rootProject.group
 version = rootProject.version
 
 dependencies {
-    implementation(platform("io.ktor:ktor-bom:2.3.10"))
-    implementation("io.ktor:ktor-client-core")
-    implementation("io.ktor:ktor-client-cio")
-    implementation("io.ktor:ktor-client-content-negotiation")
-    implementation("io.ktor:ktor-serialization-kotlinx-json")
+    // note: see https://youtrack.jetbrains.com/issue/KT-67618
```
Comment (krzema12) [inline shared-internal/build.gradle.kts:12]:
```suggestion
    // Note: BOM cannot be used here because this module is a transitive dependency
    // of a module that is used from within Kotlin scripts.
    // See https://youtrack.jetbrains.com/issue/KT-67618
```

File: shared-internal/src/main/kotlin/io/github/typesafegithub/workflows/shared/internal/GithubApi.kt [hunk 7]
```diff
@@ -29,21 +30,26 @@ private fun List<GithubRef>.versions(): List<Version> =
 
 private suspend fun fetchGithubRefs(
     url: String,
-    githubToken: String,
+    githubToken: String?,
 ): List<GithubRef> =
     httpClient.get(urlString = url) {
-        bearerAuth(githubToken)
+        expectSuccess = true
```
Comment (krzema12) [inline shared-internal/src/main/kotlin/io/github/typesafegithub/workflows/shared/internal/GithubApi.kt:36]: Is it needed (`expectSuccess = true`)? It can break other usages where where I'm trying to avoid exception-driven approach and check for the HTTP status myself.

File: shared-internal/src/main/kotlin/io/github/typesafegithub/workflows/shared/internal/GithubApi.kt [hunk 8]
```diff
@@ -29,21 +30,26 @@ private fun List<GithubRef>.versions(): List<Version> =
...
     owner: String,
     name: String,
-): String = "https://api.github.com/repos/$owner/$name/git/matching-refs/tags/v"
+): String = "https://api.github.com/repos/$owner/${repoName(name)}/git/matching-refs/tags/v"
 
 private fun apiBranchesUrl(
     owner: String,
     name: String,
-): String = "https://api.github.com/repos/$owner/$name/git/matching-refs/heads/v"
+): String = "https://api.github.com/repos/$owner/${repoName(name)}/git/matching-refs/heads/v"
```
Comment (krzema12) [inline shared-internal/src/main/kotlin/io/github/typesafegithub/workflows/shared/internal/GithubApi.kt:52]: Why are these changes needed? Perhaps to handle actions that are nested in a repo (e.g. `gradle/actions/setup-gradle`)?

I think I should better document what https://github.com/typesafegithub/github-workflows-kt/blob/fc0930030406b5769068d5283519789350d7add2/shared-internal/src/main/kotlin/io/github/typesafegithub/workflows/shared/internal/GithubApi.kt#L15
means in `fetchAvailableVersions`. I think currently the caller is supposed to path a proper repo name, without the directory name (e.g. `actions`, not `actions/setup-gradle`). Could we make it this way so that this piece remains untouched, so that we don't need to test other uses of this util?

File: shared-internal/src/main/kotlin/io/github/typesafegithub/workflows/shared/internal/PathUtils.kt [hunk 9]
```diff
@@ -7,7 +7,8 @@ import kotlin.io.path.isDirectory
 
 fun Path.findGitRoot(): Path {
     return generateSequence(this.absolute()) { it.parent }
-        .firstOrNull {
-            it.resolve(".git")?.let { it.exists() && it.isDirectory() } ?: false
+        .firstOrNull { folder ->
+            val dotGit = folder.resolve(".git")
+            dotGit.exists() && dotGit.isDirectory()
```
Comment (krzema12) [inline shared-internal/src/main/kotlin/io/github/typesafegithub/workflows/shared/internal/PathUtils.kt:12]: Why is this change needed?

File: action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/Reporting.kt [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+import io.github.typesafegithub.workflows.domain.Workflow
+import io.github.typesafegithub.workflows.shared.internal.findGitRoot
+import io.github.typesafegithub.workflows.shared.internal.getGithubTokenOrNull
+import kotlin.io.path.absolute
+import kotlin.io.path.name
+import kotlin.io.path.pathString
+import kotlin.io.path.relativeTo
+
+public fun Workflow.reportAvailableUpdates(
```
Comment (krzema12) [inline action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/Reporting.kt:11]: I see two modules exposing public functions: here `Reporting.kt` and `Utils.kt`. Could you describe what's your intended way of using the new module? I mean, we should make the API as simple as possible, perhaps hiding some implementation details as `internal`/`private`.

File: action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/model/RegularActionVersions.kt [hunk 11]
```diff
@@ -0,0 +1,12 @@
+package io.github.typesafegithub.workflows.updates.model
+
+import io.github.typesafegithub.workflows.domain.ActionStep
+import io.github.typesafegithub.workflows.domain.actions.RegularAction
+import io.github.typesafegithub.workflows.shared.internal.model.Version
+
+public data class RegularActionVersions(
```
Comment (krzema12) [inline action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/model/RegularActionVersions.kt:7]: Does it need to be public?

**Review by krzema12 — CHANGES_REQUESTED** [review]
Partial review (but review most of it 😉 ).

File: action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/GithubLogging.kt [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+            col = col,
+            endColumn = endColumn,
+            line = line,
+            endLine = endLine,
+        )
+    }
+}
+
+@OptIn(ExperimentalContracts::class)
+public fun githubGroup(
```
[the remark this answers is not in the document]
Comment (NikkyAI) [reply action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/GithubLogging.kt:161]: true, i somehow ended up not using it.. this allows you to make a nice folding out section in the logs, do we want to use that ?

File: shared-internal/src/main/kotlin/io/github/typesafegithub/workflows/shared/internal/GithubApi.kt (hunk 7 above)
Comment (NikkyAI) [reply shared-internal/src/main/kotlin/io/github/typesafegithub/workflows/shared/internal/GithubApi.kt:36]: it is not necessary but then we get back the json error and parsing fails instead in some cryptic way

File: shared-internal/src/main/kotlin/io/github/typesafegithub/workflows/shared/internal/GithubApi.kt (hunk 8 above)
Comment (NikkyAI) [reply shared-internal/src/main/kotlin/io/github/typesafegithub/workflows/shared/internal/GithubApi.kt:52]: yeah, i see now, i am moving the `name.substringBefore('/')` into the calling code instead

File: shared-internal/src/main/kotlin/io/github/typesafegithub/workflows/shared/internal/PathUtils.kt (hunk 9 above)
Comment (NikkyAI) [reply shared-internal/src/main/kotlin/io/github/typesafegithub/workflows/shared/internal/PathUtils.kt:12]: it is not, i am not sure why it slipped in.. it makes it more readable though

File: action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/model/RegularActionVersions.kt (hunk 11 above)
Comment (NikkyAI) [reply action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/model/RegularActionVersions.kt:7]: i intended it to be so people could just get a list of available versions and build their own reporting

File: action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/Reporting.kt (hunk 10 above)
Comment (NikkyAI) [reply action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/Reporting.kt:11]: i did not really spend al ong time thinking about the naming of the modules

but we have the ready to use default `reportVersionUpdates` funtion that is just one call

and the other module is exposed to give consumers the ability to list all available versions and do their own filtering and formatting

if you think we should only expose the minimum required then we can make all the rest internal

File: action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/Utils.kt [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+    }
+}
+
+internal fun RegularAction<*>.mavenCoordinatesForAction(version: Version? = null): String {
+    val mavenGroupId = actionOwner.replace("/", "__")
+    val mavenArtifactId = actionName.replace("/", "__")
+    return "$mavenGroupId:$mavenArtifactId:${version?.version ?: actionVersion}"
+}
+
+public fun Workflow.findDependencyDeclaration(action: RegularAction<*>): Pair<Path?, Int?> {
```
Comment (krzema12) [inline action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/Utils.kt:101]: How about making this internal as well? Then the API consists of a single function.

File: action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/Reporting.kt (hunk 10 above)
Comment (krzema12) [reply action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/Reporting.kt:11]: Since it will be the only exposed function, could you add some Kdoc putting most focus on what kind of side effects this function has (since it returns nothing)? Like if you were to learn someone how to use the new module, this place is probably the most important one.

Comment (NikkyAI) [reply action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/Reporting.kt:11]: will do

File: action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/Utils.kt (hunk 13 above)
Comment (NikkyAI) [reply action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/Utils.kt:101]: oh, i overlooked one.. sure can do

File: shared-internal/src/main/kotlin/io/github/typesafegithub/workflows/shared/internal/GithubApi.kt (hunk 7 above)
Comment (krzema12) [reply shared-internal/src/main/kotlin/io/github/typesafegithub/workflows/shared/internal/GithubApi.kt]: Ok, let me think about it outside of this PR, thanks for flagging!

File: action-versions-updates/build.gradle.kts [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+
+    id("org.jetbrains.kotlinx.binary-compatibility-validator") version "0.14.0"
+}
+
+group = rootProject.group
+version = rootProject.version
+
+dependencies {
+    // see https://youtrack.jetbrains.com/issue/KT-67618
+    implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core:1.8.0")
```
[the remark this answers is not in the document]
Comment (krzema12) [reply action-versions-updates/build.gradle.kts:15]: "we cannot use a BOM" - but it's just a single dependency, why would we need a BOM?

File: action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/Reporting.kt [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+import kotlin.io.path.name
+import kotlin.io.path.pathString
+import kotlin.io.path.relativeTo
+
+/**
+ * will report all available updates in the terminal output and the github step summary
+ * looks up the github token from env `GITHUB_TOKEN` by default
+ * when no github token is present and not otherwise configured, reporting will be skipped
+ *
+ * @param skipAlreadyUpToDate enabled by default, set to false to report when a version is up to date
```
Comment (krzema12) [inline action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/Reporting.kt:16]: I'm wondering if we need this param. The goal of this function is to report available updates, so why would someone want to list something more?

File: action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/Utils.kt [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+import io.github.typesafegithub.workflows.updates.model.RegularActionVersions
+import kotlinx.coroutines.runBlocking
+import java.nio.file.Path
+import kotlin.io.path.exists
+import kotlin.io.path.readText
+
+internal fun <R> Workflow.availableVersionsForEachAction(
+    reportWhenTokenUnset: Boolean = true,
+    githubToken: String? = getGithubTokenOrNull(),
+    onEach: suspend RegularActionVersions.() -> R,
```
Comment (krzema12) [inline action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/Utils.kt:18]: Instead of this param, can we make this function simply accept `RegularActionVersions`, produce a list (or a sequence if needed) and then with `map` do any processing on each?

File: action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/Utils.kt [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+): List<R> {
+    if (githubToken == null && !reportWhenTokenUnset) {
+        githubError("github token is required, but not set, skipping api calls")
+        return emptyList()
+    }
+    val groupedSteps = groupStepsByAction()
+    return runBlocking {
```
Comment (krzema12) [inline action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/Utils.kt:25]: Why is it necessary to block on it? Could we have suspending functions up to a point where the top-level function is a non-suspending one, for the caller's convenience (AFAIK the Kotlin script's entry point is not marked as `suspend`)?

File: action-versions-updates/src/test/kotlin/io/github/typesafegithub/workflows/dsl/CheckVersionsTests.kt [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+import io.github.typesafegithub.workflows.updates.findDependencyDeclaration
+import io.github.typesafegithub.workflows.updates.mavenCoordinatesForAction
+import io.github.typesafegithub.workflows.updates.reportAvailableUpdatesInternal
+import io.kotest.core.spec.style.FunSpec
+import io.kotest.engine.spec.tempdir
+import io.kotest.matchers.collections.shouldHaveSize
+import io.kotest.matchers.shouldBe
+import io.kotest.matchers.string.shouldNotBeEmpty
+
+class CheckVersionsTests : FunSpec(
```
Comment (krzema12) [inline action-versions-updates/src/test/kotlin/io/github/typesafegithub/workflows/dsl/CheckVersionsTests.kt:17]: Could you rename the file and the class so that it matches the name of the file where the tested function comes from?

File: action-versions-updates/src/test/kotlin/io/github/typesafegithub/workflows/dsl/CheckVersionsTests.kt [hunk 19]
```diff
@@ -0,0 +1,120 @@
+package io.github.typesafegithub.workflows.dsl
```
Comment (krzema12) [inline action-versions-updates/src/test/kotlin/io/github/typesafegithub/workflows/dsl/CheckVersionsTests.kt:1]:
```suggestion
package io.github.typesafegithub.workflows.updates
```

+ please move the file accordingly.

File: action-versions-updates/src/test/kotlin/io/github/typesafegithub/workflows/dsl/CheckVersionsTests.kt (hunk 18 above)
Comment (krzema12) [inline action-versions-updates/src/test/kotlin/io/github/typesafegithub/workflows/dsl/CheckVersionsTests.kt:17]: This test is failing locally if no token is passed. It's not super-convenient as one has to manually configure the token in multiple Gradle tasks. I'm wondering if we could skip running the tests if `GITHUB_TOKEN` is not set? The tests would still run on CI, and this change would give a smoother experience locally.

Comment (NikkyAI) [reply action-versions-updates/src/test/kotlin/io/github/typesafegithub/workflows/dsl/CheckVersionsTests.kt:17]: my recommendation is to configure the token in your configuration templates, that way any new task run with gradle have those defaults copied over

i am not too familar with kotest and if there is a more convenient way to make tests conditional other than `if(token != null)`

File: action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/Utils.kt (hunk 16 above)
Comment (NikkyAI) [reply action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/Utils.kt]: much cleaner, good idea

File: action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/Reporting.kt (hunk 15 above)
Comment (NikkyAI) [reply action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/Reporting.kt:16]: lets remove it then

File: action-versions-updates/src/test/kotlin/io/github/typesafegithub/workflows/dsl/CheckVersionsTests.kt (hunk 18 above)
Comment (NikkyAI) [reply action-versions-updates/src/test/kotlin/io/github/typesafegithub/workflows/dsl/CheckVersionsTests.kt]: i guess it should be named `ReportingTest` ?

File: action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/Utils.kt (hunk 17 above)
Comment (NikkyAI) [reply action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/Utils.kt]: i removed the blocking in that function and move it to the top level function
changing the return type from sequence to flow

File: action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/GithubLogging.kt (hunk 12 above)
[the remark this answers is not in the document]
Comment (NikkyAI) [reply action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/GithubLogging.kt]: see here: https://github.com/NikkyAI/github-workflows-kt/actions/runs/8816064410/job/24199345944#step:6:19

File: action-versions-updates/build.gradle.kts (hunk 14 above)
[the remark this answers is not in the document]
Comment (NikkyAI) [reply action-versions-updates/build.gradle.kts:15]: removed that comment, it was another weirdness coming from a incomplete git merge or rebase i think ?

File: action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/GithubLogging.kt (hunk 12 above)
[the remark this answers is not in the document]
Comment (krzema12) [reply action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/GithubLogging.kt]: Ok, looks nice! We could also add "click to unfold", it may be not obvious for users.

File: action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/GithubLogging.kt (hunk 4 above)
Comment (krzema12) [reply action-versions-updates/src/main/kotlin/io/github/typesafegithub/workflows/updates/GithubLogging.kt]: It's an internal function, YAGNI - we'll add once needed.

File: .github/workflows/end-to-end-tests.main.kts [hunk 20]
```diff
@@ -1,6 +1,7 @@
 #!/usr/bin/env kotlin
 @file:Repository("file://~/.m2/repository/")
 @file:DependsOn("io.github.typesafegithub:github-workflows-kt:1.14.1-SNAPSHOT")
+@file:DependsOn("io.github.typesafegithub:action-versions-updates:1.14.1-SNAPSHOT")
```
Comment (krzema12) [inline .github/workflows/end-to-end-tests.main.kts:4]: Please add another check here https://github.com/typesafegithub/github-workflows-kt/blob/c677bdfad8e3ff7f4b98d72f22a1ae73650a29b4/github-workflows-kt/build.gradle.kts#L89

Comment (NikkyAI) [reply .github/workflows/end-to-end-tests.main.kts:4]: added

File: .github/workflows/end-to-end-tests.main.kts [hunk 21]
```diff
@@ -252,5 +252,7 @@ workflow(
         )
     }
 }.also { workflow ->
-    workflow.reportAvailableUpdates()
+    if(System.getenv("GHWKT_RUN_STEP") == null) {
```
[the remark this answers is not in the document]
Comment (NikkyAI) [reply .github/workflows/end-to-end-tests.main.kts:255]: that would block people from ever using it in a kotlin logic step ?, unless e add another parameter i guess

File: action-updates-checker/build.gradle.kts [hunk 22]
```diff
@@ -0,0 +1,21 @@
+plugins {
+    buildsrc.convention.`kotlin-jvm`
+    buildsrc.convention.publishing
+    kotlin("plugin.serialization")
```
Comment (krzema12) [inline action-updates-checker/build.gradle.kts:4]: Is it needed? I think since GitHub API calls happen in shared-internal now, it should not be needed.

**Review by krzema12 — APPROVED** [review]

**Review by krzema12 — APPROVED** [review]
